Relax restrictions on multiple sanitizers

djkoloski opened the pull request:
Most combinations of LLVM sanitizers are legal-enough to enable simultaneously. This change will allow simultaneously enabling ASAN and shadow call stacks on supported platforms.

I used this python script to generate the mutually-exclusive sanitizer combinations:

```python
#!/usr/bin/python3

import subprocess

flags = [
    ["-fsanitize=address"],
    ["-fsanitize=leak"],
    ["-fsanitize=memory"],
    ["-fsanitize=thread"],
    ["-fsanitize=hwaddress"],
    ["-fsanitize=cfi", "-flto", "-fvisibility=hidden"],
    ["-fsanitize=memtag", "--target=aarch64-linux-android", "-march=armv8a+memtag"],
    ["-fsanitize=shadow-call-stack"],
    ["-fsanitize=kcfi", "-flto", "-fvisibility=hidden"],
    ["-fsanitize=kernel-address"],
    ["-fsanitize=safe-stack"],
    ["-fsanitize=dataflow"],
]

for i in range(len(flags)):
    for j in range(i):
        command = ["clang++"] + flags[i] + flags[j] + ["-o", "main.o", "-c", "main.cpp"]
        completed = subprocess.run(command, stderr=subprocess.DEVNULL)
        if completed.returncode != 0:
            first = flags[i][0][11:].replace('-', '').upper()
            second = flags[j][0][11:].replace('-', '').upper()
            print(f"(SanitizerSet::{first}, SanitizerSet::{second}),")
```

ilovepi commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1316,6 +1316,32 @@ bitflags::bitflags! {
 rustc_data_structures::external_bitflags_debug! { SanitizerSet }
 
 impl SanitizerSet {
+    const MUTUALLY_EXCLUSIVE: &'static [(SanitizerSet, SanitizerSet)] = &[
+        (SanitizerSet::MEMORY, SanitizerSet::ADDRESS),
+        (SanitizerSet::MEMORY, SanitizerSet::LEAK),
+        (SanitizerSet::THREAD, SanitizerSet::ADDRESS),
+        (SanitizerSet::THREAD, SanitizerSet::LEAK),
+        (SanitizerSet::THREAD, SanitizerSet::MEMORY),
+        (SanitizerSet::HWADDRESS, SanitizerSet::ADDRESS),
+        (SanitizerSet::HWADDRESS, SanitizerSet::MEMORY),
+        (SanitizerSet::HWADDRESS, SanitizerSet::THREAD),
+        (SanitizerSet::MEMTAG, SanitizerSet::ADDRESS),
+        (SanitizerSet::MEMTAG, SanitizerSet::HWADDRESS),
+        (SanitizerSet::KCFI, SanitizerSet::CFI),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::ADDRESS),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::LEAK),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::MEMORY),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::THREAD),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::HWADDRESS),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::MEMTAG),
+        (SanitizerSet::SAFESTACK, SanitizerSet::ADDRESS),
+        (SanitizerSet::SAFESTACK, SanitizerSet::LEAK),
+        (SanitizerSet::SAFESTACK, SanitizerSet::MEMORY),
+        (SanitizerSet::SAFESTACK, SanitizerSet::THREAD),
+        (SanitizerSet::SAFESTACK, SanitizerSet::HWADDRESS),
+        (SanitizerSet::SAFESTACK, SanitizerSet::KERNELADDRESS),
```
This doesn't quite seem right. For instance SafeStack is compatible with all sanitizers. In fact, Fuchsia uses it everywhere on X86_64 w/ ASAN. I can't recall if we fully support TSAN or KASAN yet, but it shouldn't be incompatible. 

The canonical definition in clang is here https://github.com/llvm/llvm-project/blob/70b79a9ccd03f93fc4c8464a91b6bef3aab322d3/clang/lib/Driver/SanitizerArgs.cpp#L520

djkoloski replied:
The definitions you linked seem to suggest that SafeStack is not compatible with all sanitizers:

```
std::make_pair(SanitizerKind::SafeStack,
               (TC.getTriple().isOSFuchsia() ? SanitizerMask()
                                             : SanitizerKind::Leak) |
                   SanitizerKind::Address | SanitizerKind::HWAddress |
                   SanitizerKind::Thread | SanitizerKind::Memory |
                   SanitizerKind::KernelAddress),
```

Though this change doesn't currently capture that Fuchsia specifically allows LSAN + SafeStack.

ilovepi replied:
Hmm, maybe it is ok, I realize now that we disable SafeStack w/ ASAN, so disregard. But the link can be a good cross reference.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8943540428/job/24568491871) [(plain)](https://github.com/rust-lang/rust/commit/91e927d9daf929569fa2141830064ea48b52a3b8/checks/24568491871/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.6s done
#16 DONE 40.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
diff of stderr:

1 error: cfi sanitizer is not supported for this target
2 
- error: `-Zsanitizer=cfi` is incompatible with `-Zsanitizer=kcfi`
+ error: `-Zsanitizer=kcfi` is incompatible with `-Zsanitizer=cfi`
4 
- error: `-Zsanitizer=cfi` is incompatible with `-Zsanitizer=kcfi`
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- error: aborting due to 3 previous errors
+ error: aborting due to 2 previous errors
10 
11 
11 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.aarch64/cfi-is-incompatible-with-kcfi.aarch64.stderr
To only update this specific test, also pass `--test-args sanitizer/cfi-is-incompatible-with-kcfi.rs`


error in revision `aarch64`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sanitizer/cfi-is-incompatible-with-kcfi.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "aarch64" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.aarch64" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.aarch64/auxiliary" "--target" "aarch64-unknown-none" "-Clto" "-Cno-prepopulate-passes" "-Ctarget-feature=-crt-static" "-Zsanitizer=cfi" "-Zsanitizer=kcfi"
--- stderr -------------------------------
error: cfi sanitizer is not supported for this target


error: `-Zsanitizer=kcfi` is incompatible with `-Zsanitizer=cfi`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] tests/ui/sanitizer/cfi-is-incompatible-with-kcfi.rs#x86_64 stdout ----
diff of stderr:

1 error: cfi sanitizer is not supported for this target
2 
- error: `-Zsanitizer=cfi` is incompatible with `-Zsanitizer=kcfi`
+ error: `-Zsanitizer=kcfi` is incompatible with `-Zsanitizer=cfi`
4 
- error: `-Zsanitizer=cfi` is incompatible with `-Zsanitizer=kcfi`
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- error: aborting due to 3 previous errors
+ error: aborting due to 2 previous errors
10 
11 
11 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.x86_64/cfi-is-incompatible-with-kcfi.x86_64.stderr
To only update this specific test, also pass `--test-args sanitizer/cfi-is-incompatible-with-kcfi.rs`

error in revision `x86_64`: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sanitizer/cfi-is-incompatible-with-kcfi.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "x86_64" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.x86_64" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/cfi-is-incompatible-with-kcfi.x86_64/auxiliary" "--target" "x86_64-unknown-none" "-Clto" "-Cno-prepopulate-passes" "-Ctarget-feature=-crt-static" "-Zsanitizer=cfi" "-Zsanitizer=kcfi"
--- stderr -------------------------------
error: cfi sanitizer is not supported for this target


error: `-Zsanitizer=kcfi` is incompatible with `-Zsanitizer=cfi`
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] tests/ui/sanitizer/incompatible.rs stdout ----
diff of stderr:

- error: `-Zsanitizer=address` is incompatible with `-Zsanitizer=memory`
+ error: `-Zsanitizer=memory` is incompatible with `-Zsanitizer=address`
3 error: aborting due to 1 previous error
4 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/incompatible/incompatible.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args sanitizer/incompatible.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/sanitizer/incompatible.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/incompatible" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/sanitizer/incompatible/auxiliary" "-Z" "sanitizer=address" "-Z" "sanitizer=memory" "--target" "x86_64-unknown-linux-gnu"
--- stderr -------------------------------
--- stderr -------------------------------
error: `-Zsanitizer=memory` is incompatible with `-Zsanitizer=address`
error: aborting due to 1 previous error
------------------------------------------


```

</details>

ilovepi commented:
> Since Address Sanitizer includes the leak detection, combining it with Leak Sanitizer should omit the Leak Sanitizer runtime library:
> 
> https://github.com/rust-lang/llvm-project/blob/5399a24c66cb6164cf32280e7d300488c90d5765/clang/include/clang/Driver/SanitizerArgs.h#L92-L96

That's a good point. I had assumed `rustc` already had similar logic, but I suppose if multiple sanitizers were disallowed, maybe it did not. This is some more evidence that on the LLVM side, we should probably keep that logic in a support library that frontends like `clang` and `rustc` can reuse without duplicating the same logic everywhere.

djkoloski committed:
Sort mutually-exclusive pairs, update fixed tests

djkoloski committed:
Relax restrictions on multiple sanitizers

Most combinations of LLVM sanitizers are legal-enough to enable
simultaneously. This change will allow simultaneously enabling ASAN and
shadow call stacks on supported platforms.

djkoloski committed:
Don't link lsan rt if asan or hwasan are enabled

djkoloski commented:
r? @cuviper

tmandry commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -1316,6 +1316,32 @@ bitflags::bitflags! {
 rustc_data_structures::external_bitflags_debug! { SanitizerSet }
 
 impl SanitizerSet {
+    const MUTUALLY_EXCLUSIVE: &'static [(SanitizerSet, SanitizerSet)] = &[
```
Can you add the link @ilovepi posted as a cross reference here?

djkoloski replied:
Done

rcvalle commented:
LGTM (@cuviper FYI). Thank you for your time and for working on this, @djkoloski! Much appreciated.

djkoloski committed:
Add source for mutually-exclusive list

cuviper commented:
Thanks!

@bors r=cuviper,rcvalle

bors commented:
:pushpin: Commit 1e1143c49185af1a2b3f0daa4dd3332106f51b88 has been approved by `cuviper,rcvalle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper,rcvalle 1e1143c49185af1a2b3f0daa4dd3332106f51b88 -->
<!-- homu: {"type":"Approved","sha":"1e1143c49185af1a2b3f0daa4dd3332106f51b88","approver":"cuviper,rcvalle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1e1143c49185af1a2b3f0daa4dd3332106f51b88 with merge 506512391b1a75ae450d36c9420978402a91abcc...
<!-- homu: {"type":"BuildStarted","head_sha":"1e1143c49185af1a2b3f0daa4dd3332106f51b88","merge_sha":"506512391b1a75ae450d36c9420978402a91abcc"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9177357818/job/25240018119)
Approved by: cuviper,rcvalle
Pushing 506512391b1a75ae450d36c9420978402a91abcc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper,rcvalle","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9177357818/job/25240018119"},"merge_sha":"506512391b1a75ae450d36c9420978402a91abcc"} -->

rust-timer commented:
Finished benchmarking commit (506512391b1a75ae450d36c9420978402a91abcc): [comparison URL](https://perf.rust-lang.org/compare.html?start=6715446db6564901c2674950b7fce5b487f2d13f&end=506512391b1a75ae450d36c9420978402a91abcc&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6715446db6564901c2674950b7fce5b487f2d13f&end=506512391b1a75ae450d36c9420978402a91abcc&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.4% | [1.0%, 1.7%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6715446db6564901c2674950b7fce5b487f2d13f&end=506512391b1a75ae450d36c9420978402a91abcc&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=6715446db6564901c2674950b7fce5b487f2d13f&end=506512391b1a75ae450d36c9420978402a91abcc&stat=cycles:u)
<details>
<summary>Results (primary -5.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.5% | [-5.5%, -5.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -5.5% | [-5.5%, -5.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6715446db6564901c2674950b7fce5b487f2d13f&end=506512391b1a75ae450d36c9420978402a91abcc&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.673s -> 670.767s (-0.43%)
**Artifact size**: 315.38 MiB -> 315.37 MiB (-0.00%)
